cirq/sim/act_on_density_matrix_args.py

```diff
@@ -0,0 +1,119 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Objects and methods for acting efficiently on a density matrix."""
+
+from typing import Any, Iterable, Dict, List, Tuple
+
+import numpy as np
+
+from cirq import protocols, sim
+from cirq.sim.act_on_args import ActOnArgs, strat_act_on_from_apply_decompose
+
+
+class ActOnDensityMatrixArgs(ActOnArgs):
+    """State and context for an operation acting on a density matrix.
+
+    To act on this object, directly edit the `target_tensor` property, which is
+    storing the density matrix of the quantum system with one axis per qubit.
+    """
+
+    def __init__(
+        self,
+        target_tensor: np.ndarray,
+        available_buffer: List[np.ndarray],
+        axes: Iterable[int],
+        qid_shape: Tuple[int, ...],
+        prng: np.random.RandomState,
+        log_of_measurement_results: Dict[str, Any],
+    ):
+        """
+        Args:
+            target_tensor: The state vector to act on, stored as a numpy array
+                with one dimension for each qubit in the system. Operations are
+                expected to perform inplace edits of this object.
+            available_buffer: A workspace with the same shape and dtype as
+                `target_tensor`. Used by operations that cannot be applied to
+                `target_tensor` inline, in order to avoid unnecessary
+                allocations.
+            axes: The indices of axes corresponding to the qubits that the
+                operation is supposed to act upon.
+            qid_shape: The shape of the target tensor.
+            prng: The pseudo random number generator to use for probabilistic
+                effects.
+            log_of_measurement_results: A mutable object that measurements are
+                being recorded into. Edit it easily by calling
+                `ActOnStateVectorArgs.record_measurement_result`.
+        """
+        super().__init__(axes, prng, log_of_measurement_results)
+        self.target_tensor = target_tensor
+        self.available_buffer = available_buffer
+        self.qid_shape = qid_shape
+
+    def _act_on_fallback_(self, action: Any, allow_decompose: bool):
+        strats = [
+            _strat_apply_channel_to_state,
+        ]
+        if allow_decompose:
+            strats.append(strat_act_on_from_apply_decompose)  # type: ignore
+
+        # Try each strategy, stopping if one works.
+        for strat in strats:
+            result = strat(action, self)
+            if result is False:
+                break  # coverage: ignore
+            if result is True:
+                return True
+            assert result is NotImplemented, str(result)
+        return TypeError(
```

Comment:
this could/should be a raise - that way we don't need to change anything in the act_on_protocol itself, right?

Reply:
Yes, that seems better. Thanks!